fgsm_calculations.py

```diff
@@ -0,0 +1,124 @@
+import numpy as np
+import tensorflow as tf
+import random
+import constants
+
+"""Functions used in calculation of FGSM and random perturbations"""
+
+def _get_single_perturb_dir(model_builder, answer_num, feature_num):
+    """
+    Calculates the gradient of loss with respect to a singular scalar
+    feature embedding. Used to calculate FGSM perturbation.
+
+    Args:
+      model_builder: The model in use.
+      answer_num: Integer representing answer to be perturbed.
+      feature_num: Integer for feature for which to calculate gradient with respect to.
+
+    Returns:
+      Gradient of loss with respect to singular feature embedding.
+    """
+    features = model_builder.normalized_features_evaluated
+    direction = 0
+    layer_size = constants._HIDDEN_LAYER_DIMS[0]
+    for num in range(layer_size):
+        grad = model_builder.grad_variable_pair_evaluated[0][feature_num][num]
+        weight = model_builder.grad_variable_pair_evaluated[1][feature_num][num]
+        feature = features[answer_num][feature_num]
+        # mathmatically feature value has no impact here since in coming
+        # steps we take the sign of the calculation but was included for
+        # completeness.
+        if feature != 0:

```

Comment:
Given precision issues in floats, **never** check floats for equality (or inequalities) the way you would do integers

Related: https://stackoverflow.com/questions/4682889/is-floating-point-ever-ok/4682941#4682941

instead of `a == b` ? check `abs(a - b) <= eps`?
instead of `a < b` ? check `a + eps < b `?
instead of `a <= b` ? check `a < b + eps` ?